### PR TITLE
[Enhancement] Improve DataCache io adaptor

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## What I'm doing:
1. fix "if user set invalid mem size for data cache, BE will exit immediately"
2. CacheInputStream will record network latency for each IO request, so the adaptive algorithm can be adjusted more timely.
3.  Unify duplicate function `populate_to_datacache()` in CacheInputStream
4. Move some util functions to DataCacheUtils.h

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/block_cache/datacache_utils.h [hunk 1]
```diff
@@ -22,6 +22,21 @@ namespace starrocks {
 class DataCacheUtils {
 public:
     static void set_metrics_from_thrift(TDataCacheMetrics& t_metrics, const DataCacheMetrics& metrics);
+
+    static Status parse_conf_datacache_mem_size(const std::string& conf_mem_size_str, int64_t mem_limit,
```
Comment (Smith-Cruise) [inline be/src/block_cache/datacache_utils.h:26]: just move code

File: be/src/block_cache/datacache_utils.cpp [hunk 2]
```diff
@@ -36,4 +44,112 @@ void DataCacheUtils::set_metrics_from_thrift(TDataCacheMetrics& t_metrics, const
     t_metrics.__set_mem_used_bytes(metrics.mem_used_bytes);
 }
 
+Status DataCacheUtils::parse_conf_datacache_mem_size(const std::string& conf_mem_size_str, int64_t mem_limit,
+                                                     size_t* mem_size) {
+    int64_t parsed_mem_size = ParseUtil::parse_mem_spec(conf_mem_size_str, mem_limit);
+    if (mem_limit > 0 && parsed_mem_size > mem_limit) {
+        parsed_mem_size = mem_limit;
+        LOG(WARNING) << "the configured datacache memory size exceeds the limit, decreased it to the limit value."
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/block_cache/datacache_utils.cpp:51]: done,

File: be/src/block_cache/cache_options.cpp [hunk 3]
```diff
@@ -14,116 +14,4 @@
...
-            LOG_IF(WARNING, !ret.ok()) << "fail to delete residual datacache file: " << file
-                                       << ", reason: " << ret.message();
-        }
-        return true;
-    });
-    LOG_IF(WARNING, !st.ok()) << "fail to clean residual datacache data, reason: " << st.message();
-}
-
-} // namespace starrocks
+namespace starrocks {} // namespace starrocks
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/block_cache/cache_options.cpp]: done

File: be/src/service/service_be/starrocks_be.cpp [hunk 4]
```diff
@@ -94,12 +95,13 @@ Status init_datacache(GlobalEnv* global_env, const std::vector<StorePath>& stora
                 // Clear the residual datacache files
                 std::filesystem::path sp(root_path.path);
                 auto old_path = sp.parent_path() / "datacache";
-                clean_residual_datacache(old_path.string());
+                DataCacheUtils::clean_residual_datacache(old_path.string());
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/service/service_be/starrocks_be.cpp:98]: done

File: be/src/io/cache_input_stream.cpp [hunk 5]
```diff
@@ -193,15 +171,26 @@ Status CacheInputStream::_read_blocks_from_remote(const int64_t offset, const in
...
+                const uint8_t* buffer = nullptr;
+                RETURN_IF_ERROR(_sb_stream->get_bytes(&buffer, read_offset_cursor, read_size, sb));
+                src = (char*)buffer;
+            } else {
+                RETURN_IF_ERROR(_sb_stream->read_at_fully(read_offset_cursor, _buffer.data(), read_size));
+                src = _buffer.data();
+            }
+        }
+        int64_t delta_remote_bytes =
+                _sb_stream->shared_io_bytes() + _sb_stream->direct_io_bytes() - previous_remote_bytes;
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/io/cache_input_stream.cpp]: done

File: be/src/io/cache_input_stream.cpp [hunk 6]
```diff
@@ -415,18 +373,24 @@ StatusOr<std::string_view> CacheInputStream::peek(int64_t count) {
...
-void CacheInputStream::_populate_cache_from_zero_copy_buffer(const char* p, int64_t offset, int64_t count,
-                                                             const SharedBufferPtr& sb) {
+void CacheInputStream::_populate_to_cache(const char* p, int64_t offset, int64_t count, const SharedBufferPtr& sb) {
     int64_t begin = offset / _block_size * _block_size;
     int64_t end = std::min((offset + count + _block_size - 1) / _block_size * _block_size, _size);
     p -= (offset - begin);
     auto f = [sb, this](const char* buf, size_t off, size_t size) {
+        DCHECK(off % _block_size == 0);
         SCOPED_RAW_TIMER(&_stats.write_cache_ns);
+
```
Comment (GavinMar) [inline be/src/io/cache_input_stream.cpp:388]: Move the scoped timer below after checking its existence.

Comment (Smith-Cruise) [reply be/src/io/cache_input_stream.cpp]: done

**Review by GavinMar — APPROVED** [review]

File: be/src/block_cache/CMakeLists.txt [hunk 7]
```diff
@@ -21,7 +21,7 @@ set(EXECUTABLE_OUTPUT_PATH "${BUILD_DIR}/src/block_cache")
 set(CACHE_FILES
   block_cache.cpp
   io_buffer.cpp
-  cache_options.cpp
+  cache_options.h
```
Comment (kevincai) [inline be/src/block_cache/CMakeLists.txt:24]: why include .h file in the list?

**Review by kevincai — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**